### PR TITLE
[Delta Uniform] overwrite source column field id for Iceberg PartitionField with field id assigned by Delta

### PR DESCRIPTION
**Pull request by lzlfred**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

context: Delta and Iceberg traverse schema and assigns field id in different way. Delta uniform currently use a extra Iceberg txn to overwrite the schema in iceberg table with wrong field ids reassigned by Iceberg.

However, if the source field id for partition columns is different in that schema overwrite txn, Iceberg always expect field id for partition columns to be the same and does not have logic to reconcile that, and then Iceberg will fail the overwrite txn. This PR adds the logic to adopt new source column field id to Iceberg PartitionField if changed, so the overwrite txn can go through and set the correct source column field id for PartitionFields.

## How was this patch tested?

manual test; unit test will come soon.

### CODE REVIEW

File: icebergShaded/iceberg_src_patches/0005-iceberg-takes-updated-source-column-field-id.patch [hunk 1]
```diff
@@ -0,0 +1,38 @@
+Iceberg PartitionField takes source column field id from latest schema if changed
+
+--- a/core/src/main/java/org/apache/iceberg/TableMetadata.java
++++ b/core/src/main/java/org/apache/iceberg/TableMetadata.java
+@@ -664,13 +664,21 @@ public class TableMetadata implements Serializable {
+     return new Builder(this).upgradeFormatVersion(newFormatVersion).build();
+   }
+ 
+-  private static PartitionSpec updateSpecSchema(Schema schema, PartitionSpec partitionSpec) {
++  private static PartitionSpec updateSpecSchema(Schema schema, Schema currSchema, PartitionSpec partitionSpec) {
```
Comment (LukasRupprecht) [inline icebergShaded/iceberg_src_patches/0005-iceberg-takes-updated-source-column-field-id.patch:10]:
```suggestion
+  private static PartitionSpec updateSpecSchema(Schema newSchema, Schema currSchema, PartitionSpec partitionSpec) {
```

File: icebergShaded/iceberg_src_patches/0005-iceberg-takes-updated-source-column-field-id.patch [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+         PartitionSpec.builderFor(schema).withSpecId(partitionSpec.specId());
+ 
+-    // add all the fields to the builder. IDs should not change.
++    // add all the fields to the builder. IDs may change so it looks up the source field id by
++    // name from schema to update
+     for (PartitionField field : partitionSpec.fields()) {
+-      specBuilder.add(field.sourceId(), field.fieldId(), field.name(), field.transform());
++      String partFieldSourceName = currSchema.findField(field.sourceId()).name();
++      int partFieldSourceInt;
++      if (schema.findField(partFieldSourceName) == null) {
```
Comment (LukasRupprecht) [inline icebergShaded/iceberg_src_patches/0005-iceberg-takes-updated-source-column-field-id.patch:21]: Make `schema.findField(partFieldSourceName)` a val and replace here and in line 24 below.

**Review by LukasRupprecht — APPROVED** [review]
2 nits, otherwise LGTM!

**Review by tdas — APPROVED** [review]
is this fixing a bug in uniform? then should this be backported to 3.1 branch?

**Pull request closed without merging**